nab/detectors/goodness_of_fit/goodness_of_fit_detector.py

```diff
@@ -0,0 +1,156 @@
+
+    #  This check is for files where self.inputMin == self.input max i.e
+    #  all data points are identical and stepSize is 0 e.g
+    #  artificalNoAnomaly/art_flatline.csv file. Every point in such files
+    #  is declared non-anomolous.
+    if self.stepSize != 0.0:
+
+      # All points in the first window are declared non-anomolous and
+      # anomaly detection begins when length of data points seen is
+      # greater than window length.
+      if len(self.util) >= self.W:
+
+        # Extracting current window
+        util_current = self.util[-self.W:]
+
+        # Quantize window data points into discretized bin values
+        B_current = [math.ceil((c - self.inputMin) / self.stepSize) for c in
+                     util_current]
+
+        # Create a histogram of empirial frequencies for the current window
+        # using B_current
+        P_hat = numpy.histogram(B_current,
+                                bins=self.N_bins,
+                                range=(0,self.N_bins),
+                                density=True)[0]
+
+        if self.m == 0:
+          self.P.append(P_hat)
+          self.c.append(1)
+          self.m = 1
+        else:
+          index = self.getAgreementHypothesis(P_hat)
+
+          # Check if any null hypothesis is accepted or rejected
+          if index != -1:
+
+            # If hypothesis accepted, update counter for hypothesis that tracks
+            # number of windows that have agreed to it so far.
+            self.c[index] += 1
+
+            # Check if hypothesis accepted occurs at least as frequently as
+            # the given threshold. If not, classify data point as anomolous.
+            if self.c[index] <= self.c_th:
+              anomalyScore = 1.0
+          else:
+
+            # If all null hypothesis rejected, create new hypothesis based
+            # on current window and update variables tracking hypothesis counts.
+            anomalyScore = 1.0
+            self.P.append(P_hat)
+            self.c.append(1)
+            self.m += 1
+
+    return [anomalyScore]
+
+
+  def getAgreementHypothesis(self,P_hat):
+    """This function computes multinomial goodness-of-fit test. It calculates
+    the relative entropy test statistic between P_hat and all `m` null
+    hypothesis and compares it against the threshold `T` based on cdf of
+    chi-squared distribution. The test relies on the observation that if the
+    null hypothesis P is true, then as the number of samples grow the relative
+    entropy converges to a chi-squared distribution1 with K-1 degrees of
+    freedom.
+
+    The function returns the index of hypothesis that agrees with minimum
+    relative entropy.If all hypothesis disagree,the function returns -1.
+
+    @param P_hat: empirical frequency of the current window
+    @return: index of the hypothesis with the minimum test statistic
+    """
+
+    index = -1
+    minEntropy = float("inf")
+    for i in range(0,self.m):
```

Comment:
`for i in xrange(self.m):`
Starting at 0 is implicit, and xrange is more efficient (range creates a list in memory to iterate through, whereas xrange is a sequence object that evaluates lazily).